akka-javasdk/src/main/scala/akka/javasdk/impl/agent/MemoryFiltersSupplierImpl.scala

```diff
@@ -0,0 +1,86 @@
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[javasdk] class MemoryFiltersSupplierImpl(val filters: List[MemoryFilter]) extends MemoryFilterSupplier {
+
+  def this(filter: MemoryFilter) = this(List(filter))
+
+  override def includeFromAgentId(id: String): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.IncludeFromAgentId(util.Set.of(id)))
+
+  override def includeFromAgentId(ids: util.Set[String]): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.IncludeFromAgentId(ids))
+
+  override def excludeFromAgentId(id: String): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.ExcludeFromAgentId(util.Set.of(id)))
+
+  override def excludeFromAgentId(ids: util.Set[String]): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.ExcludeFromAgentId(ids))
+
+  override def includeFromAgentRole(role: String): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.IncludeFromAgentRole(util.Set.of(role)))
+
+  override def includeFromAgentRole(roles: util.Set[String]): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.IncludeFromAgentRole(roles))
+
+  override def excludeFromAgentRole(role: String): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.ExcludeFromAgentRole(util.Set.of(role)))
+
+  override def excludeFromAgentRole(roles: util.Set[String]): MemoryFilterSupplier =
+    addFilter(new MemoryFilter.ExcludeFromAgentRole(roles))
+
+  private def addFilter(filter: MemoryFilter): MemoryFilterSupplier = {
+
+    def concatList(l1: util.Set[String], l2: util.Set[String]) = {
+      val newList = new util.HashSet(l1)
+      newList.addAll(l2)
+      newList
+    }
+
+    val newFilters =
+      if (filters.exists(_.getClass == filter.getClass)) {
+        filters
+          .map {
+            case f: MemoryFilter.IncludeFromAgentId if f.getClass == filter.getClass =>
+              val ids = filter.asInstanceOf[MemoryFilter.IncludeFromAgentId].ids()
+              new MemoryFilter.IncludeFromAgentId(concatList(f.ids(), ids))
+
+            case f: MemoryFilter.ExcludeFromAgentId if f.getClass == filter.getClass =>
+              val ids = filter.asInstanceOf[MemoryFilter.ExcludeFromAgentId].ids()
+              new MemoryFilter.ExcludeFromAgentId(concatList(f.ids(), ids))
+
+            case f: MemoryFilter.IncludeFromAgentRole if f.getClass == filter.getClass =>
+              val roles = filter.asInstanceOf[MemoryFilter.IncludeFromAgentRole].roles()
+              new MemoryFilter.IncludeFromAgentRole(concatList(f.roles(), roles))
+
+            case f: MemoryFilter.ExcludeFromAgentRole if f.getClass == filter.getClass =>
+              val roles = filter.asInstanceOf[MemoryFilter.ExcludeFromAgentRole].roles()
+              new MemoryFilter.ExcludeFromAgentRole(concatList(f.roles(), roles))
+
+            case any => any // making compiler happy
+          }
+      } else {
+        filters :+ filter // Append the new filter to preserve existing filters
+      }
```

Comment:
We need to merge the filters for the same kind. For example, all include-by-id must be concatenated otherwise, when applying them one by one, we will filter out prematurely. 

Also, in a hindsight, combining include-by-id and exclude-by-id makes no sense. Same for mixing include/exclude by role. But I'm leaving it as is, because making such API will require lots of work and low yield at the end.